lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/classify/MultiMADClassifier.java

```diff
@@ -0,0 +1,291 @@
+    private List<String> columnNames;
+    private String outputColumnSuffix = "_OUTLIER";
+    private boolean useParallel = false;
+    private boolean useReservoirSampling = false;
+    private final double trimmedMeanFallback = 0.05;
+    // https://en.wikipedia.org/wiki/Median_absolute_deviation#Relation_to_standard_deviation
+    private final double MAD_TO_ZSCORE_COEFFICIENT = 1.4826;
+
+    // Calculated values
+    private DataFrame output;
+
+    public MultiMADClassifier(String... columnNames) {
+        this.columnNames = new ArrayList<String>(Arrays.asList(columnNames));
+    }
+
+    @Override
+    public void process(DataFrame input) {
+        output = input.copy();
+
+        int sampleSize = (int)(input.getNumRows() * samplingRate);
+        int[] sampleIndices = new int[sampleSize];
+        if (samplingRate != 1) {
+            sampleIndices = getRandomSample(input.getNumRows());
+        }
+
+        for (String columnName : columnNames) {
+            double[] metrics = input.getDoubleColumnByName(columnName);
+
+            double[] trainingMetrics = new double[sampleSize];
+            if (samplingRate == 1) {
+                System.arraycopy(metrics, 0, trainingMetrics, 0, metrics.length);
+            } else {
+                for (int j = 0; j < sampleIndices.length; j++) {
+                    trainingMetrics[j] = metrics[sampleIndices[j]];
+                }
+            }
+
+            Stats stats = train(trainingMetrics);
+            double lowCutoff = stats.median - (zscore * stats.mad);
+            double highCutoff = stats.median + (zscore * stats.mad);
+
+            if (useParallel) {
+                // Parallelized stream scoring:
+                double[] results = new double[metrics.length];
+                System.arraycopy(metrics, 0, results, 0, metrics.length);
+                Arrays.parallelSetAll(results, m -> (
+                    (m > highCutoff && includeHigh) || (m < lowCutoff && includeLow)) ? 1.0 : 0.0);
+                output.addDoubleColumn(columnName + outputColumnSuffix, results);
+            } else {
+                // Non-parallel scoring:
+                double[] results = new double[metrics.length];
+                for (int r = 0; r < input.getNumRows(); r++) {
+                    boolean isOutlier = (metrics[r] > highCutoff && includeHigh) ||
+                        (metrics[r] < lowCutoff && includeLow);
+                    results[r] = isOutlier ? 1.0 : 0.0;
+                }
+                output.addDoubleColumn(columnName + outputColumnSuffix, results);
+            }
+        }
+    }
+
+    private int[] getRandomSample(int numRows) {
+        int sampleSize = (int)(numRows * samplingRate);
+        int[] sampleIndices = new int[sampleSize];
+
+        if (useReservoirSampling) {
+            // Reservoir Sampling:
+            for (int i = 0; i < sampleSize; i++) {
+                sampleIndices[i] = i;
+            }
+            Random rand = new Random();
```

Comment:
We should factor out all the "Random" generators used, ideally share them within the same class. That way we can optionally set seeds in the constructor for deterministic behavior